Reflection_Engine/Query/ClassHierarchy.cs

```diff
@@ -0,0 +1,49 @@
+﻿/*
+ * This file is part of the Buildings and Habitats object Model (BHoM)
+ * Copyright (c) 2015 - 2018, the respective contributors. All rights reserved.
+ *
+ * Each contributor holds copyright over their respective contributions.
+ * The project versioning (Git) records all such contribution source information.
+ *                                           
+ *                                                                              
+ * The BHoM is free software: you can redistribute it and/or modify         
+ * it under the terms of the GNU Lesser General Public License as published by  
+ * the Free Software Foundation, either version 3.0 of the License, or          
+ * (at your option) any later version.                                          
+ *                                                                              
+ * The BHoM is distributed in the hope that it will be useful,              
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of               
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the                 
+ * GNU Lesser General Public License for more details.                          
+ *                                                                            
+ * You should have received a copy of the GNU Lesser General Public License     
+ * along with this code. If not, see <https://www.gnu.org/licenses/lgpl-3.0.html>.      
+ */
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace BH.Engine.Reflection
+{
+    public static partial class Query
+    {
+        /***************************************************/
+        /**** Public Methods                            ****/
+        /***************************************************/
+
+        public static List<Type> ClassHierarchy(this Type type)
+        {
+            if (type == null)
+                return new List<Type>();
+
+            List<Type> classHierarchy = new List<Type>();
+            for (Type current = type; current != null; current = current.GetInterfaces().FirstOrDefault() ?? current.BaseType)
```

Comment:
The `FirstOrDefault` is incorrect here. You need to collect all of them (plus base type) and then apply `ClassHierarchy` recursively on those for it to give the correct result.

Reply:
I found `InheritedTypes` doing exactly what I need, with a small fix. I will delete `ClassHierarchy`